File: Lab2/r/solution.R
euclidean_distance <- function(x1, y1, x2, y2) {
  sqrt((x1 - x2)^2 + (y1 - y2)^2)
}

find_nearest_centroid <- function(px, py, centroids) {
  dists <- mapply(function(cx, cy) euclidean_distance(px, py, cx, cy),
                  centroids$x, centroids$y)
  which.min(dists)
}

init_centroids <- function(points, m, seed = 42) {
  if (nrow(points) < m) stop("Число точек меньше числа кластеров")
  set.seed(seed)
  idx <- sample(nrow(points), m)
  data.frame(x = points$x[idx], y = points$y[idx])
}

assign_clusters <- function(points, centroids) {
  sapply(seq_len(nrow(points)), function(i) {
    find_nearest_centroid(points$x[i], points$y[i], centroids)
  })
}

update_centroids <- function(points, assignments, m) {
  centroids <- data.frame(x = numeric(m), y = numeric(m))
  for (c in seq_len(m)) {
    mask <- assignments == c
    if (any(mask)) {
      centroids$x[c] <- mean(points$x[mask])
      centroids$y[c] <- mean(points$y[mask])
    }
  }
  centroids
}

determine_cluster_class <- function(points, assignments, cluster_id) {
  labels <- points$label[assignments == cluster_id]
  if (length(labels) == 0) return(-1L)
  tbl <- table(labels)
  as.integer(names(which.max(tbl)))
}

compute_total_distance <- function(points, assignments, centroids) {
  total <- 0.0
  for (i in seq_len(nrow(points))) {
    c <- assignments[i]
    total <- total + euclidean_distance(points$x[i], points$y[i],
                                        centroids$x[c], centroids$y[c])
  }
  total
}

kmeans_custom <- function(points, m, max_iter = 300, tol = 1e-6, seed = 42) {
  if (nrow(points) == 0) stop("Массив точек пуст")
  if (m <= 0) stop("Число кластеров должно быть > 0")

  centroids <- init_centroids(points, m, seed)
  assignments <- integer(nrow(points))

  iter <- 0
  for (it in seq_len(max_iter)) {
    iter <- it
    assignments <- assign_clusters(points, centroids)
    new_centroids <- update_centroids(points, assignments, m)

    shift <- sum(mapply(function(i) {
      euclidean_distance(centroids$x[i], centroids$y[i],
                         new_centroids$x[i], new_centroids$y[i])
    }, seq_len(m)))

    centroids <- new_centroids
    if (shift < tol) break
  }

  cluster_info <- list()
  total_mismatched <- 0L

  for (c in seq_len(m)) {
    assigned_class <- determine_cluster_class(points, assignments, c)
    mask <- assignments == c
    total_pts <- sum(mask)
    mismatched <- sum(points$label[mask] != assigned_class)

    cluster_info[[c]] <- list(
      cluster_id = c,
      assigned_class = assigned_class,
      total_points = total_pts,
      mismatched_points = mismatched
    )
    total_mismatched <- total_mismatched + mismatched
  }

  error_rate <- total_mismatched / nrow(points)
  total_dist <- compute_total_distance(points, assignments, centroids)

  list(
    assignments = assignments,
    centroids = centroids,
    cluster_info = cluster_info,
    error_rate = error_rate,
    total_distance = total_dist,
    iterations = iter
  )
}

generate_points <- function(n, k, seed = 42) {
  set.seed(seed)
  centers_x <- runif(k, -50, 50)
  centers_y <- runif(k, -50, 50)

  labels <- sample(0:(k - 1), n, replace = TRUE)
  x <- centers_x[labels + 1] + rnorm(n, 0, 5)
  y <- centers_y[labels + 1] + rnorm(n, 0, 5)

  data.frame(x = x, y = y, label = labels)
}

print_results <- function(result, points) {
  cat("\n=== Результаты кластеризации ===\n")
  cat(sprintf("Итераций:            %d\n", result$iterations))
  cat(sprintf("Суммарное расстояние: %.2f\n", result$total_distance))
  cat(sprintf("Доля ошибок:         %.4f%%\n", result$error_rate * 100))

  cat("\nКластеры:\n")
  cat(sprintf("%10s%10s%10s%12s%14s%14s\n",
              "Кластер", "Класс", "Точек", "Ошибок", "Центр X", "Центр Y"))
  cat(strrep("-", 70), "\n")
  for (ci in result$cluster_info) {
    cat(sprintf("%10d%10d%10d%12d%14.4f%14.4f\n",
                ci$cluster_id, ci$assigned_class, ci$total_points,
                ci$mismatched_points,
                result$centroids$x[ci$cluster_id],
                result$centroids$y[ci$cluster_id]))
  }

  show <- min(nrow(points), 20)
  cat(sprintf("\nПервые %d точек:\n", show))
  cat(sprintf("%8s%12s%12s%10s%12s\n", "#", "X", "Y", "Класс", "Кластер"))
  cat(strrep("-", 54), "\n")
  for (i in seq_len(show)) {
    cat(sprintf("%8d%12.4f%12.4f%10d%12d\n",
                i - 1, points$x[i], points$y[i], points$label[i],
                result$assignments[i]))
  }
}

main <- function() {
  cat("Число точек (n): ")
  n <- as.integer(readLines(con = "stdin", n = 1))

  cat("Число классов (k): ")
  k <- as.integer(readLines(con = "stdin", n = 1))

  cat("Число кластеров (m): ")
  m <- as.integer(readLines(con = "stdin", n = 1))

  cat("Ввод данных (gen/manual): ")
  mode <- trimws(readLines(con = "stdin", n = 1))

  if (mode == "gen") {
    cat(sprintf("\nГенерация %d точек...\n", n))
    points <- generate_points(n, k)
  } else {
    cat(sprintf("Введите %d троек (x y class):\n", n))
    x <- numeric(n); y <- numeric(n); label <- integer(n)
    for (i in seq_len(n)) {
      parts <- as.numeric(strsplit(readLines(con = "stdin", n = 1), " ")[[1]])
      x[i] <- parts[1]; y[i] <- parts[2]; label[i] <- as.integer(parts[3])
    }
    points <- data.frame(x = x, y = y, label = label)
  }

  cat("Кластеризация...\n")
  result <- kmeans_custom(points, m)
  print_results(result, points)
}

if (!interactive() && !exists(".IS_TEST")) {
  args <- commandArgs(trailingOnly = TRUE)
  if (length(args) == 0 || args[1] != "--test") {
    main()
  }
}

File: Lab2/python/solution.py
import math
import random
from collections import Counter


def euclidean_distance(x1, y1, x2, y2):
    return math.sqrt((x1 - x2) ** 2 + (y1 - y2) ** 2)


def find_nearest_centroid(point, centroids):
    best = 0
    best_dist = float("inf")
    for i, c in enumerate(centroids):
        d = euclidean_distance(point[0], point[1], c[0], c[1])
        if d < best_dist:
            best_dist = d
            best = i
    return best


def init_centroids(points, m, seed=42):
    if len(points) < m:
        raise ValueError("Число точек меньше числа кластеров")
    rng = random.Random(seed)
    indices = rng.sample(range(len(points)), m)
    return [(points[i][0], points[i][1]) for i in indices]


def assign_clusters(points, centroids):
    return [find_nearest_centroid(p, centroids) for p in points]


def update_centroids(points, assignments, m):
    sums = [[0.0, 0.0, 0] for _ in range(m)]
    for i, p in enumerate(points):
        c = assignments[i]
        sums[c][0] += p[0]
        sums[c][1] += p[1]
        sums[c][2] += 1

    centroids = []
    for s in sums:
        if s[2] > 0:
            centroids.append((s[0] / s[2], s[1] / s[2]))
        else:
            centroids.append((0.0, 0.0))
    return centroids


def determine_cluster_class(points, assignments, cluster_id):
    labels = [p[2] for i, p in enumerate(points) if assignments[i] == cluster_id]
    if not labels:
        return -1
    counter = Counter(labels)
    return counter.most_common(1)[0][0]


def compute_total_distance(points, assignments, centroids):
    total = 0.0
    for i, p in enumerate(points):
        c = centroids[assignments[i]]
        total += euclidean_distance(p[0], p[1], c[0], c[1])
    return total


def kmeans(points, m, max_iter=300, tol=1e-6, seed=42):
    if not points:
        raise ValueError("Массив точек пуст")
    if m <= 0:
        raise ValueError("Число кластеров должно быть > 0")

    centroids = init_centroids(points, m, seed)
    assignments = []

    for iteration in range(1, max_iter + 1):
        assignments = assign_clusters(points, centroids)
        new_centroids = update_centroids(points, assignments, m)

        shift = sum(
            euclidean_distance(c[0], c[1], nc[0], nc[1])
            for c, nc in zip(centroids, new_centroids)
        )
        centroids = new_centroids

        if shift < tol:
            break

    cluster_info = []
    total_mismatched = 0

    for c in range(m):
        assigned_class = determine_cluster_class(points, assignments, c)
        total = 0
        mismatched = 0
        for i, p in enumerate(points):
            if assignments[i] == c:
                total += 1
                if p[2] != assigned_class:
                    mismatched += 1
        cluster_info.append({
            "cluster_id": c,
            "assigned_class": assigned_class,
            "total_points": total,
            "mismatched_points": mismatched,
        })
        total_mismatched += mismatched

    error_rate = total_mismatched / len(points)
    total_dist = compute_total_distance(points, assignments, centroids)

    return {
        "assignments": assignments,
        "centroids": centroids,
        "cluster_info": cluster_info,
        "error_rate": error_rate,
        "total_distance": total_dist,
        "iterations": iteration,
    }


def generate_points(n, k, seed=42):
    rng = random.Random(seed)
    centers = [(rng.uniform(-50, 50), rng.uniform(-50, 50)) for _ in range(k)]

    points = []
    for _ in range(n):
        label = rng.randint(0, k - 1)
        x = centers[label][0] + rng.gauss(0, 5)
        y = centers[label][1] + rng.gauss(0, 5)
        points.append((x, y, label))
    return points


def print_results(result, points):
    print(f"\n=== Результаты кластеризации ===")
    print(f"Итераций:            {result['iterations']}")
    print(f"Суммарное расстояние: {result['total_distance']:.2f}")
    print(f"Доля ошибок:         {result['error_rate'] * 100:.4f}%")

    print(f"\nКластеры:")
    print(f"{'Кластер':>10}{'Класс':>10}{'Точек':>10}{'Ошибок':>12}{'Центр X':>14}{'Центр Y':>14}")
    print("-" * 70)
    for ci in result["cluster_info"]:
        cid = ci["cluster_id"]
        cx, cy = result["centroids"][cid]
        print(f"{cid:>10}{ci['assigned_class']:>10}{ci['total_points']:>10}"
              f"{ci['mismatched_points']:>12}{cx:>14.4f}{cy:>14.4f}")

    show = min(len(points), 20)
    print(f"\nПервые {show} точек:")
    print(f"{'#':>8}{'X':>12}{'Y':>12}{'Класс':>10}{'Кластер':>12}")
    print("-" * 54)
    for i in range(show):
        p = points[i]
        print(f"{i:>8}{p[0]:>12.4f}{p[1]:>12.4f}{p[2]:>10}{result['assignments'][i]:>12}")


def main():
    n = int(input("Число точек (n): "))
    k = int(input("Число классов (k): "))
    m = int(input("Число кластеров (m): "))
    mode = input("Ввод данных (gen/manual): ").strip()

    if mode == "gen":
        print(f"\nГенерация {n} точек...")
        points = generate_points(n, k)
    else:
        print(f"Введите {n} троек (x y class):")
        points = []
        for _ in range(n):
            parts = input().split()
            points.append((float(parts[0]), float(parts[1]), int(parts[2])))

    print("Кластеризация...")
    result = kmeans(points, m)
    print_results(result, points)


if __name__ == "__main__":
    main()
